ast.go

```diff
@@ -0,0 +1,274 @@
+package mf2
+
+import (
+	"golang.org/x/exp/constraints"
+)
+
+// AST is the abstract syntax tree of a MessageFormat 2.0 source file.
+type AST Message
+
+// --------------------------------Interfaces----------------------------------
+//
+// Here we define the Nodes that can have multiple types.
+// For example Message could be either a SimpleMessage or a ComplexMessage.
+// Pattern could be either a TextPattern or a PlaceholderPattern.
+// etc.
+
+// Node is the interface implemented by all AST nodes.
+type Node interface {
+	node()
+}
+
+// Message is the top-level node
+type Message interface {
+	Node
+	message()
+}
+
+type Pattern interface {
+	Node
+	pattern()
+}
+
+type Expression interface {
+	Node
+	expression()
+}
+
+type Literal interface {
+	Node
+	literal()
+}
+
+type Unquoted interface {
+	Literal
+	unquoted()
+}
+
+type Annotation interface {
+	Node
+	annotation()
+}
+
+type Option interface {
+	Node
+	option()
+}
+
+type Declaration interface {
+	Node
+	declaration()
+}
+
+type ComplexBody interface {
+	Node
+	complexBody()
+}
+
+type VariantKey interface {
+	Node
+	variantKey()
+}
+
+// ---------------------------------Structs------------------------------------
+//
+// Here we define the structs that implement the interfaces defined above.
+
+// ---------------------------------Message------------------------------------
+
+type SimpleMessage struct {
+	Message
+
+	Patterns []Pattern // TextPattern or PlaceholderPattern
+}
+
+type ComplexMessage struct {
+	Message
+
+	Declarations []Declaration // Optional: InputDeclaration, LocalDeclaration or ReservedStatement
+	ComplexBody  ComplexBody   // Matcher or QuotedPattern
+}
+
+// ---------------------------------Pattern------------------------------------
+
+type TextPattern struct {
+	Pattern
+
+	Text string
+}
+
+type PlaceholderPattern struct {
+	Pattern
+
+	Expression Expression // LiteralExpression, VariableExpression, or AnnotationExpression
+}
+
+// --------------------------------Expression----------------------------------
+
+type LiteralExpression struct {
```

Comment:
Can we simplify and have only one `Expression` struct? Expression is annotation with optional (variable | literal).

Reply:
We could, but I would rather not.
**Reason**: I want our `AST` tree to look like `ABNF` syntax as much as possible, and combining in one expression compromises that.